### PR TITLE
Introduce RequireAmazonClientBuildItem for dependent extensions

### PR DESCRIPTION
**Pull request by scrocquesel**
Refs: https://github.com/quarkiverse/quarkus-amazon-services/issues/458

The PR allow to register the enhanced client only if needed.
It does so by expressing a requirement for a client to wrap and set the bean instance lazily with a bean container listener, so that the client extension do not register the client because the wrapper producer has injection point of it.

It may be possible to do it another way but that's the only one that come to my mind. SyntheticBeanBuildItem are package protected and its RuntimeValue cannot be reuse further.

I'm also not sure about the `@Record(ExecutionTime.STATIC_INIT)` on the `DynamodbEnhancedProcessor.createClientBuilders` while client is recorded for RUNTIME_INIT.

@gsmet Can you have a look at it ?

### CODE REVIEW

File: dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import io.quarkus.arc.runtime.BeanContainerListener;
+import io.quarkus.runtime.annotations.Recorder;
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+
+@Recorder
+public class DynamodbEnhancedClientRecorder {
+
+    public BeanContainerListener setDynamoDbClient() {
+        return beanContainer -> {
```
Comment (gsmet) [inline dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java:12]: Let's avoid lambdas in runtime code. They have a cost. It's fine for processors but we try to limit them in runtime code to places where they really are needed.

File: dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java [hunk 2]
```diff
@@ -43,6 +66,58 @@ AdditionalBeanBuildItem producer() {
...
+            }
+        }
+
+        if (syncClassName.isPresent() || asyncClassName.isPresent()) {
+            requireClientProducer.produce(new RequireAmazonClientBuildItem(syncClassName, asyncClassName));
+        }
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
```
Comment (gsmet) [inline dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java:96]: I think it's fine given what the methods recorded do.

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks! See inline!

File: dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java (hunk 1 above)
Comment (scrocquesel) [reply dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientRecorder.java]: done

File: dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java (hunk 2 above)
Comment (scrocquesel) [reply dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java:96]: integration test passed and a user testing in real application reported it worked without regression, so should be safe

**Review by gsmet — APPROVED** [review]

**Pull request merged**